polaris-react/src/components/IndexTable/IndexTable.stories.tsx

```diff
@@ -1347,6 +1347,122 @@ export function WithAllOfItsElements() {
   }
 }
 
+export function WithSortableHeadings() {
+  const [sortIndex, setSortIndex] = useState(0);
+  const [sortDirection, setSortDirection] = useState('descending');
+
+  const initialRows = [
+    {
+      id: '3411',
+      url: 'customers/341',
+      name: 'Mae Jemison',
+      location: 'Decatur, USA',
+      orders: 20,
+      amountSpent: '$2,400',
+    },
+    {
+      id: '2561',
+      url: 'customers/256',
+      name: 'Ellen Ochoa',
+      location: 'Los Angeles, USA',
+      orders: 30,
+      amountSpent: '$140',
+    },
+    {
+      id: '1245',
+      url: 'customers/123',
+      name: 'Anne-Marie Johnson',
+      location: 'Portland, USA',
+      orders: 10,
+      amountSpent: '$250',
+    },
+    {
+      id: '8741',
+      url: 'customers/543',
+      name: 'Bradley Stevens',
+      location: 'Hialeah, USA',
+      orders: 5,
+      amountSpent: '$26',
+    },
+  ];
+  const [sortedRows, setSortedRows] = useState(
+    sortRows(initialRows, sortIndex, sortDirection),
+  );
+
+  const resourceName = {
+    singular: 'customer',
+    plural: 'customers',
+  };
+
+  const rows = sortedRows ?? initialRows;
+
+  const {selectedResources, allResourcesSelected, handleSelectionChange} =
+    useIndexResourceState(rows);
+
+  function handleClickSortHeading(index, direction) {
+    setSortIndex(index);
```

Comment:
If we are setting the `sortIndex` only here then passing the `index` from the `onSort` callback directly to our sort function, do we still need to keep `sortIndex` in state?

Similarly, I think the same could be done for the `setDirection`. Matching the `DataTabel` prop interface seems more consistent and less state management on the user:

```tsx
interface IndexTableBaseProps {
 ...

  /**
   * The direction to sort the table rows on first click or keypress of a sortable column heading. Defaults to ascending.
   * @default 'ascending'
   */
  defaultSortDirection?: SortDirection;
  /**
   * The index of the heading that the table rows are initially sorted by. Defaults to the first column.
   * @default 0
   */
  initialSortColumnIndex?: number;
  ...
}

Reply:
When using IndexTables in the admin, they are usually paired with some form of filter/sort controls, which means that the sorting state still needs to live outside of the IndexTable itself. Extremely simplified version below:

```tsx
function MyComponent() {
  const [sortIndex, setSortIndex] = useState(0);
  const [sortDirection, setSortDirection] = useState('ascending');

  function handleSort(index, direction) {
    setSortIndex(index)
    setSortDirection(direction)
  }

  return (
    <>
      <MyCustomSortingComponent onSort={handleSort} sortIndex={sortIndex} sortDirection={sortDirection} />

      <IndexTable sortColumnIndex={sortIndex} sortDirection={sortDirection} onSort={handleSort} {...otherProps} />
    </>
  )
}
```

We have access to the new index and direction due to the `onSort` handler providing us the new values, but we need to store them at the comsumer level as the `MyCustomSortingComponent` also needs to know these values.

If we have something like this which would bring the sorting state inside of the IndexTable:

```tsx
function MyInternalComponent() {

  function handleSort(index, direction) {
    console.log(index);
    console.log(direction);
  }

  return (
    <>
      <MyCustomSortingComponent onSort={handleSort} sortIndex="??" sortDirection="??" />

      <IndexTable defaultSortColumnIndex={0} sortDirection="ascending" onSort={handleSort} {...otherProps} />
    </>
  )
}
```

Then we do have access to the new values thanks to the `onSort` function, but given we also need to pass these values to the `MyCustomSortingComponent`, we can't access them without saving them at the consumer level.

Looking through the `web` codebase, in the examples I can find where a DataTable is sortable, there is no additional external components which can sort the data inside the table, so it makes sense to internalise the state in that case. But the IndexTable sorting implementations within `web` differ. Whether down the line we want to homogenise this is probably a broader discussion, and one that I'm sure the Quality Crew would love to be a part of if and when that time comes 😄

Reply:
I see, thank you for the added example and clarity! Makes total sense 👍 this API looks good to me